dali/pipeline/operators/arithmetic/expression_impl_factory.h

```diff
@@ -0,0 +1,97 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_FACTORY_H_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_FACTORY_H_
+
+#include "dali/core/any.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/util/backend2workspace_map.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+template <typename Backend>
+class ExpressionImplParam {
+ protected:
+  // We differentiate between TensorVector nd TensorList, that still have inconsistent
+  // access patterns
+  static constexpr bool is_cpu = std::is_same<Backend, CPUBackend>::value;
+
+  template <bool IsTensor, typename Type>
+  std::enable_if_t<IsTensor && is_cpu, const Type *> ObtainInput(const ExpressionBase &expr,
+                                                                 workspace_t<Backend> &ws,
+                                                                 const OpSpec &spec,
+                                                                 SampleExtent tile,
+                                                                 int subexpr_id) {
+    int input_id = expr[subexpr_id].GetMappedInput();
+    auto *tensor = ws.template InputRef<Backend>(input_id)[tile.sample_idx].template data<Type>();
+    return tensor + tile.extent_idx * tile.extent_size;
+  }
+
+  template <bool IsTensor, typename Type>
+  std::enable_if_t<IsTensor && !is_cpu, const Type *> ObtainInput(const ExpressionBase &expr,
+                                                                  workspace_t<Backend> &ws,
+                                                                  const OpSpec &spec,
+                                                                  SampleExtent tile,
+                                                                  int subexpr_id) {
+    int input_id = expr[subexpr_id].GetMappedInput();
+    auto *tensor = ws.template InputRef<Backend>(input_id).template tensor<Type>(tile.sample_idx);
+    return tensor + tile.extent_idx * tile.extent_size;
+  }
+
+  template <bool IsTensor, typename Type>
+  std::enable_if_t<!IsTensor, Type> ObtainInput(const ExpressionBase &expr,
+                                                workspace_t<Backend> &ws, const OpSpec &spec,
+                                                SampleExtent tile, int subexpr_id) {
+    int scalar_id = expr[subexpr_id].GetMappedInput();
+    if (expr.type_id < DALIDataType::DALI_FLOAT16) {
```

Comment:
I don't like this conditional. Consider having something like `is_float_type` or so

Reply:
done